File: apps/android/app/src/main/java/com/rayact/engine/RayactMobileNetwork.kt
package com.rayact.engine

import android.os.StrictMode
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import org.json.JSONArray
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.io.ByteArrayOutputStream
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

object RayactMobileNetwork {
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    // Separate client for JS `fetch()`: bounded timeouts (unlike the WebSocket
    // client's infinite read) and, crucially, all requests run on OkHttp's
    // dispatcher threads — never on the render/JS thread. The old synchronous
    // fetchTextFromNative/fetchBytesFromNative shims blocked the render thread
    // for up to the connect timeout, which froze the dev launcher whenever it
    // probed unreachable recent servers.
    private val fetchClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(5, TimeUnit.SECONDS)
        .readTimeout(10, TimeUnit.SECONDS)
        .callTimeout(15, TimeUnit.SECONDS)
        .build()
    private val nextSocketId = AtomicInteger(1)
    private data class SocketRecord(val owner: Long, val socket: WebSocket)
    private val sockets = ConcurrentHashMap<Int, SocketRecord>()
    private val events = ConcurrentHashMap<Long, ConcurrentLinkedQueue<String>>()

    @JvmStatic
    fun fetchTextFromNative(url: String): String {
        val previous = StrictMode.getThreadPolicy()
        StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder(previous).permitAll().build())
        return try {
            val conn = (URL(url).openConnection() as HttpURLConnection).apply {
                connectTimeout = 15_000
                readTimeout = 60_000
                instanceFollowRedirects = true
                requestMethod = "GET"
            }
            try {
                val code = conn.responseCode
                val stream = if (code in 200..299) conn.inputStream else conn.errorStream
                stream?.bufferedReader()?.use { it.readText() }.orEmpty()
            } finally {
                conn.disconnect()
            }
        } catch (_: Exception) {
            ""
        } finally {
            StrictMode.setThreadPolicy(previous)
        }
    }

    @JvmStatic
    fun fetchBytesFromNative(url: String): ByteArray {
        val previous = StrictMode.getThreadPolicy()
        StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder(previous).permitAll().build())
        return try {
            val conn = (URL(url).openConnection() as HttpURLConnection).apply {
                connectTimeout = 15_000
                readTimeout = 60_000
                instanceFollowRedirects = true
                requestMethod = "GET"
            }
            try {
                val code = conn.responseCode
                val stream = if (code in 200..299) conn.inputStream else conn.errorStream
                stream?.use { input ->
                    val out = ByteArrayOutputStream()
                    input.copyTo(out)
                    out.toByteArray()
                } ?: ByteArray(0)
            } finally {
                conn.disconnect()
            }
        } catch (_: Exception) {
            ByteArray(0)
        } finally {
            StrictMode.setThreadPolicy(previous)
        }
    }

    // Asynchronous fetch backing the JS `fetch()` polyfill. Returns immediately;
    // the response (or error) is delivered later via the per-owner event queue
    // and drained on the render thread, exactly like WebSocket events. Never
    // touches the render/JS thread with network I/O.
    @JvmStatic
    fun fetchStart(owner: Long, requestId: Int, url: String) {
        fetchCandidate(owner, requestId, devServerFetchCandidates(url), 0)
    }

    private fun fetchCandidate(owner: Long, requestId: Int, candidates: List<String>, index: Int) {
        val url = candidates[index]
        val request = try {
            Request.Builder().url(url).build()
        } catch (e: Exception) {
            enqueueFetch(owner, requestId, 0, "", e.message ?: "Invalid URL")
            return
        }
        fetchClient.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (index + 1 < candidates.size) {
                    fetchCandidate(owner, requestId, candidates, index + 1)
                    return
                }
                enqueueFetch(owner, requestId, 0, "", e.message ?: "Network request failed")
            }

            override fun onResponse(call: Call, response: Response) {
                response.use {
                    val responseBody = it.body
                    val mimeType = responseBody?.contentType()?.toString().orEmpty()
                    val bytes = responseBody?.bytes() ?: ByteArray(0)
                    // Map each byte to a char 0..255 so the body round-trips
                    // through JSON unchanged; the JS side rebuilds the exact byte
                    // array via charCodeAt & 0xff (and UTF-8-decodes for text()).
                    val sb = StringBuilder(bytes.size)
                    for (b in bytes) sb.append((b.toInt() and 0xff).toChar())
                    enqueueFetch(
                        owner, requestId, it.code, sb.toString(), "",
                        it.message, headersJson(it), mimeType, it.protocol.toString()
                    )
                }
            }
        })
    }

    /** Prefer the adb-reversed route for Rayact endpoints, with LAN fallback. */
    private fun devServerFetchCandidates(rawUrl: String): List<String> {
        val loopback = runCatching {
            val parsed = URL(rawUrl)
            if (!parsed.protocol.equals("http", ignoreCase = true) ||
                !parsed.path.startsWith("/rayact/") ||
                parsed.host.equals("localhost", ignoreCase = true) ||
                parsed.host == "127.0.0.1" || parsed.host == "::1") {
                return@runCatching null
            }
            val port = if (parsed.port >= 0) parsed.port else parsed.defaultPort
            URL(parsed.protocol, "127.0.0.1", port, parsed.file).toString()
        }.getOrNull()
        return if (loopback.isNullOrEmpty() || loopback == rawUrl) listOf(rawUrl)
        else listOf(loopback, rawUrl)
    }

    private fun enqueueFetch(
        owner: Long,
        requestId: Int,
        status: Int,
        body: String,
        error: String,
        statusText: String = "",
        headers: JSONObject = JSONObject(),
        mimeType: String = "",
        protocol: String = ""
    ) {
        val obj = JSONObject()
            .put("type", "fetch")
            .put("req", requestId)
            .put("status", status)
            .put("body", body)
            .put("statusText", statusText)
            .put("headers", headers)
            .put("mimeType", mimeType)
            .put("protocol", protocol)
        if (error.isNotEmpty()) obj.put("error", error)
        events.computeIfAbsent(owner) { ConcurrentLinkedQueue() }.add(obj.toString())
        wakeRenderFrame(owner)
    }

    @JvmStatic
    fun wsOpen(owner: Long, url: String): Int {
        val id = nextSocketId.getAndIncrement()
        val request = Request.Builder().url(url).build()
        val socket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                enqueue(owner, id, "open") {
                    put("status", response.code)
                    put("statusText", response.message)
                    put("headers", headersJson(response))
                }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                enqueue(owner, id, "message") { put("data", text) }
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                enqueue(owner, id, "message") {
                    put("data", bytes.base64())
                    put("binary", true)
                }
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                enqueue(owner, id, "close") {
                    put("code", code)
                    put("reason", reason)
                }
                sockets.remove(id)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                enqueue(owner, id, "close") {
                    put("code", code)
                    put("reason", reason)
                }
                sockets.remove(id)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                enqueue(owner, id, "error") { put("message", t.message ?: "WebSocket failure") }
                sockets.remove(id)
            }
        })
        sockets[id] = SocketRecord(owner, socket)
        return id
    }

    @JvmStatic
    fun wsSend(owner: Long, id: Int, data: String): Boolean = sockets[id]?.takeIf { it.owner == owner }?.socket?.send(data) == true

    @JvmStatic
    fun wsClose(owner: Long, id: Int, code: Int, reason: String): Boolean {
        val record = sockets[id]?.takeIf { it.owner == owner } ?: return false
        sockets.remove(id, record)
        return record.socket.close(code, reason)
    }

    @JvmStatic
    fun wsPollEvents(owner: Long): String {
        val arr = JSONArray()
        val queue = events[owner] ?: return arr.toString()
        while (true) {
            val ev = queue.poll() ?: break
            arr.put(JSONObject(ev))
        }
        return arr.toString()
    }

    // Wake the render loop from this OkHttp callback thread so an idle app
    // schedules a frame to drain the event (drained per-frame via
    // __rayactNativeNetworkDrain). Without this, incoming WebSocket messages —
    // including Chrome DevTools (CDP) traffic over /rayact/debugger — would only
    // be delivered on the next unrelated frame (up to the ~5s poll interval).
    @JvmStatic external fun nativeWakeRenderFrame(owner: Long)

    private fun wakeRenderFrame(owner: Long) {
        try {
            nativeWakeRenderFrame(owner)
        } catch (_: UnsatisfiedLinkError) {
            // Native engine not loaded yet; the periodic frame will still drain.
        }
    }

    private inline fun enqueue(owner: Long, id: Int, type: String, block: JSONObject.() -> Unit = {}) {
        val obj = JSONObject()
            .put("id", id)
            .put("type", type)
        obj.block()
        events.computeIfAbsent(owner) { ConcurrentLinkedQueue() }.add(obj.toString())
        wakeRenderFrame(owner)
    }

    @JvmStatic fun closeAll(owner: Long) {
        sockets.entries.filter { it.value.owner == owner }.forEach { (id, record) ->
            if (sockets.remove(id, record)) record.socket.cancel()
        }
        events.remove(owner)
    }

    private fun headersJson(response: Response): JSONObject = JSONObject().also { json ->
        for (name in response.headers.names()) {
            json.put(name, response.headers.values(name).joinToString("\n"))
        }
    }
}

File: apps/ios/RayactMobileNetwork.swift
import Foundation

enum RayactMobileNetwork {
    private static let lock = NSLock()
    private static var fetchStorage: [CChar] = [0]
    private static var pollStorage: [CChar] = Array("[]".utf8CString)
    private static var nextSocketId: Int32 = 1
    private struct SocketRecord { let owner: Int64; let task: URLSessionWebSocketTask }
    private static var sockets: [Int32: SocketRecord] = [:]
    private static var events: [Int64: [[String: Any]]] = [:]

    static let fetchTextCallback: RayactNativeBridge.NetworkFetchFn = { urlPtr in
        let url = urlPtr.map { String(cString: $0) } ?? ""
        let text = (try? RayactHTTP.getText(url)) ?? ""
        lock.lock()
        fetchStorage = Array(text.utf8CString)
        let ptr = fetchStorage.withUnsafeBufferPointer { $0.baseAddress }
        lock.unlock()
        return ptr
    }

    private static var networkFetchBytesStorage: [UInt8] = []

    static let fetchBytesCallback: RayactNativeBridge.NetworkFetchBytesFn = { urlPtr, outLenPtr in
        let url = urlPtr.map { String(cString: $0) } ?? ""
        let data = (try? RayactHTTP.getBytes(url)) ?? Data()
        lock.lock()
        networkFetchBytesStorage = Array(data)
        outLenPtr?.pointee = UInt32(networkFetchBytesStorage.count)
        let ptr = networkFetchBytesStorage.withUnsafeBufferPointer { $0.baseAddress }
        lock.unlock()
        return ptr
    }

    // Asynchronous fetch backing the JS `fetch()` polyfill. Returns immediately;
    // the response is delivered later via the per-owner event queue and drained
    // on the render thread. Replaces the old synchronous httpGetText/Bytes shims
    // that blocked the render thread (froze the dev launcher on unreachable
    // server probes).
    static let fetchStartCallback: RayactNativeBridge.NetworkFetchStartFn = { owner, requestId, urlPtr in
        let urlString = urlPtr.map { String(cString: $0) } ?? ""
        guard let url = URL(string: urlString) else {
            enqueueFetch(owner: owner, requestId: requestId, status: 0, body: "", error: "Invalid URL")
            return
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = 5
        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error {
                enqueueFetch(owner: owner, requestId: requestId, status: 0, body: "", error: error.localizedDescription)
                return
            }
            let http = response as? HTTPURLResponse
            let status = http?.statusCode ?? 200
            let statusText = HTTPURLResponse.localizedString(forStatusCode: status)
            let headers = (http?.allHeaderFields ?? [:]).reduce(into: [String: String]()) { out, entry in
                out[String(describing: entry.key)] = String(describing: entry.value)
            }
            let mimeType = response?.mimeType ?? ""
            // Map each byte to a Unicode scalar 0..255 so the body round-trips
            // through JSON unchanged; the JS side rebuilds the exact byte array
            // via charCodeAt & 0xff (and UTF-8-decodes for text()).
            var body = ""
            if let data {
                body.reserveCapacity(data.count)
                for b in data { body.unicodeScalars.append(UnicodeScalar(b)) }
            }
            enqueueFetch(
                owner: owner, requestId: requestId, status: status, body: body, error: "",
                statusText: statusText, headers: headers, mimeType: mimeType
            )
        }.resume()
    }

    private static func enqueueFetch(
        owner: Int64,
        requestId: Int32,
        status: Int,
        body: String,
        error: String,
        statusText: String = "",
        headers: [String: String] = [:],
        mimeType: String = ""
    ) {
        lock.lock()
        var ev: [String: Any] = [
            "type": "fetch", "req": Int(requestId), "status": status, "body": body,
            "statusText": statusText, "headers": headers, "mimeType": mimeType, "protocol": ""
        ]
        if !error.isEmpty { ev["error"] = error }
        events[owner, default: []].append(ev)
        lock.unlock()
        RayactNativeBridge.requestGraphicsFrame(owner)
    }

    static let wsOpenCallback: RayactNativeBridge.WebSocketOpenFn = { owner, urlPtr in
        let urlString = urlPtr.map { String(cString: $0) } ?? ""
        guard let url = URL(string: urlString) else { return 0 }
        lock.lock()
        let id = nextSocketId
        nextSocketId += 1
        lock.unlock()
        let task = URLSession.shared.webSocketTask(with: url)
        lock.lock()
        sockets[id] = SocketRecord(owner: owner, task: task)
        lock.unlock()
        task.resume()
        enqueue(owner: owner, id: id, type: "open")
        listen(owner: owner, id: id, task: task)
        return id
    }

    static let wsSendCallback: RayactNativeBridge.WebSocketSendFn = { owner, id, dataPtr in
        let text = dataPtr.map { String(cString: $0) } ?? ""
        lock.lock()
        let task = sockets[id].flatMap { $0.owner == owner ? $0.task : nil }
        lock.unlock()
        guard let task else { return false }
        task.send(.string(text)) { error in
            if let error { enqueue(owner: owner, id: id, type: "error", extra: ["message": error.localizedDescription]) }
        }
        return true
    }

    static let wsCloseCallback: RayactNativeBridge.WebSocketCloseFn = { owner, id, code, reasonPtr in
        let reason = reasonPtr.map { String(cString: $0) } ?? ""
        lock.lock()
        let record = sockets[id]
        let task = record?.owner == owner ? sockets.removeValue(forKey: id)?.task : nil
        lock.unlock()
        guard let task else { return false }
        task.cancel(with: URLSessionWebSocketTask.CloseCode(rawValue: Int(code)) ?? .normalClosure,
                    reason: reason.data(using: .utf8))
        enqueue(owner: owner, id: id, type: "close", extra: ["code": Int(code), "reason": reason])
        return true
    }

    static let wsPollCallback: RayactNativeBridge.WebSocketPollFn = { owner in
        lock.lock()
        let batch = events.removeValue(forKey: owner) ?? []
        let data = (try? JSONSerialization.data(withJSONObject: batch)) ?? Data("[]".utf8)
        pollStorage = Array((String(data: data, encoding: .utf8) ?? "[]").utf8CString)
        let ptr = pollStorage.withUnsafeBufferPointer { $0.baseAddress }
        lock.unlock()
        return ptr
    }

    private static func listen(owner: Int64, id: Int32, task: URLSessionWebSocketTask) {
        task.receive { result in
            switch result {
            case .success(.string(let text)):
                enqueue(owner: owner, id: id, type: "message", extra: ["data": text])
                listen(owner: owner, id: id, task: task)
            case .success(.data(let data)):
                enqueue(owner: owner, id: id, type: "message", extra: [
                    "data": data.base64EncodedString(), "binary": true
                ])
                listen(owner: owner, id: id, task: task)
            case .success:
                listen(owner: owner, id: id, task: task)
            case .failure(let error):
                enqueue(owner: owner, id: id, type: "error", extra: ["message": error.localizedDescription])
                lock.lock()
                sockets.removeValue(forKey: id)
                lock.unlock()
            }
        }
    }

    private static func enqueue(owner: Int64, id: Int32, type: String, extra: [String: Any] = [:]) {
        lock.lock()
        var ev: [String: Any] = ["id": Int(id), "type": type]
        for (k, v) in extra { ev[k] = v }
        events[owner, default: []].append(ev)
        lock.unlock()
        RayactNativeBridge.requestGraphicsFrame(owner)
    }

    static func closeAll(owner: Int64) {
        lock.lock()
        let owned = sockets.filter { $0.value.owner == owner }
        for (id, _) in owned { sockets.removeValue(forKey: id) }
        events.removeValue(forKey: owner)
        lock.unlock()
        for (_, record) in owned { record.task.cancel(with: .goingAway, reason: nil) }
    }
}
